File: BOJ/[9251] LCS/ONE/9251.py
a = input()
b = input()
x = len(a)
y = len(b)
table = [[0]*(y+1) for _ in range(x+1)]

for i in range(1, x+1):
    for j in range(1, y+1):
        if a[i-1] == b[j-1]:
           table[i][j] = table[i-1][j-1] + 1
        else:
            table[i][j] = max(table[i][j-1], table[i-1][j])
print(table[x][y])

File: BOJ/[9251] LCS/pinkishincoloragain/9251.js
const fs = require("fs");

let [a, b] = fs
	// .readFileSync("./input.txt")
	.readFileSync("./dev/stdin")
	.toString()
	.split("\n");

const dp = Array.from({ length: 1001 }, () => Array(1001).fill(0));

for (let i = 0; i < a.length; i++) {
	for (let j = 0; j < b.length; j++) {
		if (a[i] === b[j]) {
			dp[i + 1][j + 1] = dp[i][j] + 1;
		} else {
			dp[i + 1][j + 1] = Math.max(dp[i + 1][j], dp[i][j + 1]);
		}
	}
}

console.log(dp[a.length][b.length]);
